src/kernel/bitcoinkernel.cpp

```diff
@@ -1,10 +1,1410 @@
+}
+
+void kernel_byte_array_destroy(kernel_ByteArray* byte_array)
+{
+    delete[] byte_array->data;
+    delete byte_array;
+}
+
+kernel_ByteArray* kernel_copy_block_data(kernel_Block* block_, kernel_Error* error)
+{
+    auto block{cast_cblocksharedpointer(block_, error)};
+    if (!block) {
+        return nullptr;
+    }
+
+    DataStream ss{};
+    ss << TX_WITH_WITNESS(**block);
+
+    auto byte_array{new kernel_ByteArray{
+        .data = new unsigned char[ss.size()],
+        .size = ss.size(),
+    }};
+
+    std::memcpy(byte_array->data, ss.data(), byte_array->size);
+
+    return byte_array;
+}
+
+kernel_ByteArray* kernel_copy_block_pointer_data(const kernel_BlockPointer* block_, kernel_Error* error)
+{
+    auto block{cast_cblockpointer(block_, error)};
+    if (!block) {
+        return nullptr;
+    }
+
+    DataStream ss{};
+    ss << TX_WITH_WITNESS(*block);
+
+    auto byte_array{new kernel_ByteArray{
+        .data = new unsigned char[ss.size()],
+        .size = ss.size(),
+    }};
+
+    std::memcpy(byte_array->data, ss.data(), byte_array->size);
+
+    return byte_array;
+}
+
+void kernel_block_destroy(kernel_Block* block)
+{
+    delete reinterpret_cast<std::shared_ptr<CBlock>*>(block);
+}
+
+kernel_BlockIndex* kernel_get_block_index_from_tip(const kernel_Context* context_, kernel_ChainstateManager* chainman_, kernel_Error* error)
+{
+    auto context{cast_const_context(context_, error)};
+    if (!context) {
+        return nullptr;
+    }
+    auto chainman{cast_chainstate_manager(chainman_, error)};
+    if (!chainman) {
+        return nullptr;
+    }
+
+    return reinterpret_cast<kernel_BlockIndex*>(WITH_LOCK(::cs_main, return chainman->ActiveChain().Tip()));
+}
+
+kernel_BlockIndex* kernel_get_block_index_from_genesis(const kernel_Context* context_, kernel_ChainstateManager* chainman_, kernel_Error* error)
+{
+    auto context{cast_const_context(context_, error)};
+    if (!context) {
+        return nullptr;
+    }
+
+    auto chainman{cast_chainstate_manager(chainman_, error)};
+    if (!chainman) {
+        return nullptr;
+    }
+
+    return reinterpret_cast<kernel_BlockIndex*>(WITH_LOCK(::cs_main, return chainman->ActiveChain().Genesis()));
+}
+
+kernel_BlockIndex* kernel_get_block_index_by_hash(const kernel_Context* context_, kernel_ChainstateManager* chainman_, kernel_BlockHash* block_hash, kernel_Error* error)
+{
+    auto context{cast_const_context(context_, error)};
+    if (!context) {
+        return nullptr;
+    }
+
+    auto chainman{cast_chainstate_manager(chainman_, error)};
+    if (!chainman) {
+        return nullptr;
+    }
+
+    auto hash = uint256{Span<const unsigned char>{(*block_hash).hash, 32}};
+    auto block_index = WITH_LOCK(::cs_main, return chainman->m_blockman.LookupBlockIndex(hash));
+    if (!block_index) {
+        set_error(error, kernel_ErrorCode::kernel_ERROR_INTERNAL, "A block with the given hash is not indexed.");
```

Comment:
I don't think this should be an error. Just returning a nullptr here (without error message) makes for a much easier interface imo.